src/utils.c

```diff
@@ -67,3 +71,96 @@ void rv_clock_gettime(struct timespec *tp)
     tp->tv_sec = tv_sec;
     tp->tv_nsec = tv_usec / 1000; /* Transfer to microseconds */
 }
+
+char *sanitize_path(const char *input)
+{
+    size_t n = strnlen(input, MAX_PATH_LEN);
+
+    char *ret = malloc(MAX_PATH_LEN + 1);
+    memset(ret, '\0', MAX_PATH_LEN + 1);
+
+    /* After sanitization, the new path will only be shorter than the original
+     * one. Thus, we can reuse the space */
+    if (n == 0) {
+        ret[0] = '.';
+        return ret;
+    }
+
+    int rooted = (input[0] == '/');
+
+    /*
+     * Invariants:
+     * reading from path; r is index of next byte to process -> path[r]
+     * writing to buf; w is index of next byte to write -> ret[strlen(ret)]
+     * dotdot is index in buf where .. must stop, either because:
+     *  a) it is the leading slash
+     *  b) it is a leading ../../.. prefix.
+     */
+    size_t w = 0;
+    size_t r = 0;
+    size_t dotdot = 0;
+    if (rooted) {
+        ret[w] = '/';
+        w++;
+        r = 1;
+        dotdot = 1;
+    }
+
+    while (r < n) {
+        if (input[r] == '/') {
+            /*  empty path element */
+            r++;
+        } else if (input[r] == '.' && (r + 1 == n || input[r + 1] == '/')) {
+            /* . element */
+            r++;
+        } else if (input[r] == '.' && input[r + 1] == '.' &&
+                   (r + 2 == n || input[r + 2] == '/')) {
+            /* .. element: remove to last / */
+            r += 2;
+
+            if (w > dotdot) {
+                /* can backtrack */
```

Comment:
should be "backtrace"?

Reply:
It's taken from the Golang codebase. 

I believe the original meaning was indeed backtracking, as we are moving backward on the path string from the current position of `w` (e.g. `/a/b/...` will be backtracked to `a`). 

Backtrace is a stack trace, probably not what we want here IMO.